packages/meditrak-server/src/database/SurveyResponseChangeHandler/OutdatedStatusUpdater.js

```diff
@@ -0,0 +1,261 @@
+ * the rest should be marked as outdated. This one response is selected on the basis of most
+ * recent submission time
+ *
+ * Here, we refer to the survey/entity/period combo as "response dimension combo".
+ * Example: Basic Clinic Survey/Tonga/July 2021
+ *
+ * @typedef { surveyId, entityId, startDate, endDate } ResponseDimensionCombo
+ */
+
+const DATE_FORMAT = 'YYYY-MM-DD';
+const DATETIME_FORMAT = `${DATE_FORMAT} HH:mm:ss`;
+
+/**
+ * Cache of period ranges keyed by requested period granularity first, then date
+ *
+ * ```js
+ * {
+ *   yearly: {
+ *     "2021-05-05": { startDate: '2021-01-01', endDate: '2021-12-31' }
+ *   }
+ *   monthly: {
+ *     "2021-07-06": { startDate: '2021-07-01', endDate: '2021-07-31' }
+ *   }
+ * }
+ * ```
+ */
+const periodRangeCache = Object.fromEntries(
+  Object.values(PERIOD_GRANULARITIES).map(granularity => [granularity, {}]),
+);
+
+/**
+ * @param {Date|string} datetime
+ * @returns {string}
+ */
+const extractDateString = datetime => {
+  const datetimeString = typeof datetime === 'string' ? datetime : datetime.toISOString();
+  return datetimeString.substring(0, DATE_FORMAT.length);
+};
+
+/**
+ */
+const getPeriodRange = (periodGranularity, datetime) => {
+  // None of the existing granularities depends on time, so we can just use the date part
+  // to cache/retrieve date ranges
+  const date = extractDateString(datetime);
+  const existingRange = periodRangeCache?.[periodGranularity]?.[date];
+  if (existingRange) {
+    return existingRange;
+  }
+
+  const momentUnit = PERIOD_GRANULARITY_TO_MOMENT_UNIT[periodGranularity];
+  const range = {
+    startDate: moment(datetime).startOf(momentUnit).format(DATETIME_FORMAT),
+    endDate: moment(datetime).endOf(momentUnit).format(DATETIME_FORMAT),
+  };
+  periodRangeCache[periodGranularity][date] = range;
+
+  return range;
+};
+
+const getIdOfMostRecentResponse = surveyResponses => {
+  assert.notStrictEqual(surveyResponses.length, 0);
+
+  const comparableResponses = surveyResponses.map(response => ({
+    ...response,
+    // If other ordering criteria are the same, not outdated responses take precedence over outdated
+    // Assign numeric value for comparison
+    notOutdated: response.outdated ? 0 : 1,
+  }));
+
+  return orderBy(
+    comparableResponses,
+    ['end_time', 'notOutdated', 'id'],
+    ['desc', 'desc', 'desc'],
+  )[0].id;
+};
+
+const getDimensionCombo = (surveyResponse, survey) => {
+  const { period_granularity: periodGranularity } = survey;
+  const { data_time: dataTime } = surveyResponse;
+  const { startDate, endDate } = getPeriodRange(periodGranularity, dataTime);
+
+  return {
+    surveyId: survey.id,
+    entityId: surveyResponse.entity_id,
+    startDate,
+    endDate,
+  };
+};
+
+const groupResponsesForSurveyByDimensionCombo = (surveyResponses, survey) =>
+  Object.values(groupBy(surveyResponses, sr => JSON.stringify(getDimensionCombo(sr, survey))));
```

Comment:
In this PR, the object is instantiated in the same way every time: 

https://github.com/beyondessential/tupaia/blob/76fdebf396088af5bbc7af45a7f51429ec6ddc31/packages/meditrak-server/src/database/SurveyResponseChangeHandler/OutdatedStatusUpdater.js#L108

and is stringified in the same way every time (with `JSON.stringify`. So, even if the object key order does not match the insertion order, I would assume the resulting string is deterministic?

By the way, ES6 **does** guarantee the insertion order for non-number keys: https://2ality.com/2015/10/property-traversal-order-es6.html. (Using number keys in a hash is a probably a bad practice)

We don't set a [target](https://babeljs.io/docs/en/babel-preset-env#no-targets) in Babel so we currently transpile to ES5 (default), so I guess that the above doesn't apply in our case. Having said that, it's probably a good idea to change this since our server runs node `v12.18.3`!

Reply:
(I'm happy to use a 3rd party library btw, that's often the way to go, but I don't think it's required in this case)